xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,407 @@
+            elif self.single_block != op.parent_block():
+                self.used_in_single_block = False
+
+        self.users.append(op)
+
+    @staticmethod
+    def from_declare(op: symref.Declare) -> 'SymbolInfo':
+        return SymbolInfo(op)
+
+    @staticmethod
+    def from_fetch_or_update(op: symref.Fetch | symref.Update) -> 'SymbolInfo':
+        info = SymbolInfo(None)
+        info.add_user(op)
+        return info
+
+
+@dataclass
+class Desymrefier:
+    """
+    Class responsible for rewriting xDSL and removing symref operations.
+    """
+
+    rewriter: Rewriter
+    """Rewriter to replace and erase operations."""
+
+    def run_on_operation(self, op: Operation):
+        """Desymrefies an operation."""
+
+        # For operation with no regions we don't have to do any work.
+        if len(op.regions) == 0:
+            return
+
+        # Otherwise, there is a region containing a CFG and we have to desymrefy
+        # it.
+        for region in op.regions:
+            self.run_on_region(region)
+
+        # Some regions were not fully desymrefied, so use the definition of the
+        # operation to decide what to do.
+        Promoter(op).promote()
+
+    def run_on_region(self, region: Region):
+        """Desymrefies a region."""
+        num_blocks = len(region.blocks)
+        if num_blocks == 1:
+            # If there is only one block, desymrefication is significantly
+            # easier.
+            self._run_on_single_block(region.blocks[0])
+        else:
+            # TODO: Support regions with multiple blocks. This is not trivial,
+            # particularly when the symbol is declared in one of the parent
+            # regions.
+            raise FrontendProgramException(
+                f"Running desymrefier on region with {num_blocks} > 1 blocks is "
+                "not supported.")
+
+    def _run_on_single_block(self, block: Block):
+        """Desymrefies a single block inside a region."""
+
+        # First, we desymrefy nested regions.
+        for op in block.ops:
+            self.run_on_operation(op)
+
+        # Case 1: symbol is declared in this region. Then,
+        # iterate until all symbols are destroyed.
+        self._remove_declared_symbols(block)
+
+        # Case 2: some symbols are not declared in this region.
+        # For a single block, it is not that different!
+        self._remove_used_symbols(block)
+
+        # Sanity check.
+        self._check_single_block_for_promotion(block)
+
+    def _remove_declared_symbols(self, block: Block):
+        """Removes all symbol declarations in a single block."""
+        while True:
+            # Get all symbol declarations in this block.
+            declare_ops: List[symref.Declare] = []
+            for op in block.ops:
+                if isinstance(op, symref.Declare):
+                    declare_ops.append(op)
+
+            # No declarations - we are done.
+            if len(declare_ops) == 0:
```

Comment:
This will not really work though here, because this is a do-while loop. I am not sure there is any nice way to express them in Python? Otherwise we repeat the code to populate `declare_ops`

Reply:
Yes I would do the repetition approach for that. Or maybe move the repeated part to a function? Then it's not quite as annoying...